### ISSUE TITLE
Doesn't consistently stop miner processes from starting 

### ISSUE THREAD

**Issue by Leondthurrell** [keraf/NoCoin#30]
I have found that, while it blocks communication to the miner host, this plugin currently doesn't always stop the miner from trying to run. The miners will still use some processing power when they do start.

Running the `miner.sop()` command built into Coinhive's miner stops the miner if it is running.
After stopping the miner you could clear it's variable by setting it to be undefined to prevent it from being able to be started again.

```javascript
miner.stop();
miner = undefined;
```

**Comment by Leondthurrell** [comment]
Update on this. It looks like you have to set `CoinHive` to be undefined.

```javascript
miner.stop();
CoinHive = undefined;
```

Also this solution assumes that the  functions are named the same. It would be best to look into a more robust solution.

**Comment by keraf** [comment]
Thanks for reporting this, I'm going to have a closer look at that one. I'll update this issue with my findings.

**Comment by Leondthurrell** [comment]
As it is the plugin catches it most of the time but sometimes it seems the script loads before it's blocked. 
Might have to do on if the script is cached or not?

**Comment by keraf** [comment]
Thanks for taking time to look into this, it is really helpful! I am wondering though if the request watcher can intercept cached file if I would add `coinhive.min.js` to the blacklist. But I guess overriding the method is a more effective way of stopping that. 

I will experiment a bit and look into adding a content script to stop the miner as you described. This will probably be part of the next update (0.4).

**Comment by keraf** [comment]
I gave it another thought. Overriding `CoinHive` is easy enough with a content script. It can be done at different times: Before the coinhive script, which wouldn't make sense as the script (if cached) will set `CoinHive` back. Or after the coinhive script but then the start method might have already been called and to stop it I would need to now the name of the variable (which might not always be `miner`) and then override `CoinHive` to avoid having the page start the miner again.

I guess the best way to make sure to kill the miner would be to listen for new service workers and terminate them if they have a coin hive task. I think Chrome's new `chrome.process` API might be helpful for that but it's not there yet: https://developer.chrome.com/extensions/processes

For now, I guess the best way is to make sure miner scripts from the blacklist are not getting cached. Looking at the Chrome docs, I think there's a way to do that.

I'm quite tired tonight, so I made have said non-sense above or forgotten something, please let me know if it is the case. Also, if you have any solution for that problem, feel free to submit a PR or just share a PoC here. Cheers!

**Comment by Leondthurrell** [comment]
I'll keep taking a look at this. I agree that we should have a more universal solution for it for if/when a site doesn't call their miner `miner` or have a simply exposed stop method like this one currently has.

Ideally we should be able to just check for the cached files and remove them or keep them from running but I have had little luck with that direction so far.

**Comment by jspenguin2017** [comment]
`Object.defineProperty` can be used to define a `CoinHive` variable that is not overridable. But it could be a bit difficult when you want to allow it to mine, because there isn't a way to configure the behavior of content scripts synchronously. 

A more complicated way is to have a setter that saves the pointer to the real `CoinHive`, but feed back a fake `CoinHive` when the getter is called, the fake `CoinHive` will store API calls and wait for popup page to send "mine" command (relay the communication with content script), and dump the API calls backlog onto the real `CoinHive`. You wouldn't need to reload the page to start mining this way, but is a LOT more complicated as you need to implement a full `CoinHive` API so stuff don't crash. 
Here is somewhat of an example that may help, I only had to implement 1 function but you'll need to do it for a full library with all the call chaining, can be quite challenging: https://github.com/jspenguin2017/InlineInstallBlocker/blob/master/Extension/content.js

Another way, which should be the best way to go, is to alter the response header to force a revalidate every time.

**Comment by Leondthurrell** [comment]
I agree that the response header route is probably the best way to go. If we could set it up so that when the website is white listed it gets tracked files using `Cache-Control: no-cache, no-store, must-revalidate` in the header that should prevent the file from ever getting cached.

The second "more complicated" option sounds promising as well. The biggest problem I have with that route personally, other than time to implement, is how specific it is to one miner. Having an option that works with any miner and regardless of the object name would be best. Then we would just have to keep the blacklist updated versus having to modify the object name and/or the API based on every coin miner that presents itself.

**Comment by Leondthurrell** [comment]
I have tested this idea using the extension [super cache](https://chrome.google.com/webstore/detail/super-cache/fglobbnbihckpkodmeefhagijjcjnbeh?hl=en) Setting the Cache-Control to `no-store` works exactly as expected.

**Comment by keraf** [comment]
Thanks for digging into this, some useful information here, I really appreciate. I am quite busy at the moment but I will try to get onto this as soon as I can. Meanwhile, if any of you want to give it a shot, feel free to submit a PR.

**Comment by lesander** [comment]
Setting for example `const CoinHive = undefined` in the injected script might prevent the miner script from starting a new miner.
